src/state/opamGlobalState.ml

```diff
@@ -44,7 +44,12 @@ let load lock_kind =
   let has_root = OpamFilename.exists_dir root in
   let global_lock =
     if has_root then
-      OpamFilename.flock `Lock_read (OpamPath.lock root)
+      let lock = (* needed for on-the-fly upgrade config file *)
+        match lock_kind with
+        | `Lock_none | `Lock_read -> `Lock_read
```

Comment:
Wow, aren't you forcing a read-lock for all commands here ?
Why would you need this ?

Reply:
No it's only the `global_lock` that is used for the upgrade, the _real_ conf lock is [`config_lock`](https://github.com/ocaml/opam/pull/4314/files/1ddb5b1d3994cca66fdea87b09d9a1ea7f6c8b8f#diff-927760da3264767b18aaab24a8795803L56)

Reply:
ah right, of course